Adds editor options overloads to showTextDocument & vscode.diff

eamodio opened the pull request:
This PR hopefully addresses https://github.com/Microsoft/vscode/issues/10568
Also addresses a need I have with `vscode.diff` to have the same control (for GitLens)

@jrieken This is a first shot at this -- please let me know what you think. I added it as options so that it could be expanded later to allow for other editor options (but didn't want to go there at this stage).

msftclas commented:
__@eamodio__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

jrieken commented:
Check, in the middle of endgame this week. We have resources to jump on this starting next week

eamodio commented:
OK, thanks. I'll dig into the broken tests soon as well

eamodio commented:
@jrieken I resynced this with master and ran the test locally and I don't get any failures. Hopefully the builds will succeed now.

eamodio commented:
I see the failures now (in the integration tests), but I'm having trouble figuring out how to debug them. Any help would be appreciated.

jrieken commented on src/vs/vscode.d.ts:
```diff
@@ -3646,6 +3646,18 @@ declare module 'vscode' {
+		 * @param options Editor options for showing the text editor.
+		 * @return A promise that resolves to an [editor](#TextEditor).
+		 */
+		export function showTextDocument(document: TextDocument, column?: ViewColumn, options?: { preserveFocus?: boolean, pinned?: boolean }): Thenable<TextEditor>;
```
Sooo, if we add this new overload it should include `column` and also make it a type which we can document nicer

eamodio replied:
OK - Will do. Should I also allow any other editor options come through? I have it explicitly locked to just that set right now

jrieken commented on src/vs/workbench/api/node/extHost.protocol.ts:
```diff
@@ -129,7 +129,10 @@ export abstract class MainThreadDocumentsShape {
+	$tryShowTextDocument(resource: URI, position: EditorPosition, preserveFocus: boolean): TPromise<string>;
+	$tryShowTextDocument(resource: URI, position: EditorPosition, options: { preserveFocus: boolean, pinned: boolean }): TPromise<string>;
+	$tryShowTextDocument(resource: URI, position: EditorPosition, preserveFocusOrOptions: boolean | { preserveFocus: boolean, pinned: boolean }): TPromise<string>;
+	$tryShowTextDocument(resource: URI, position: EditorPosition, preserveFocusOrOptions: boolean | { preserveFocus: boolean, pinned: boolean }): TPromise<string> { throw ni(); }
```
In the protocol/shape there should only be one signature and the calling side should do the argument construction

eamodio replied:
Will do.

jrieken commented:
Please also rebase this on master to see if those tests still fail, they shouldn't. Then I wonder if we actually tackle  #10568 here. There is also an ask to (1) know if an editor is pinned or not pinned and (2) an ask to change that after the fact. We should think about exposing this information on the editor, similar to `viewColumn`, and about allowing to change that property, similar to changing a selection. However, important is to know that there are embedded and other editors that cannot be pinned etc. Last, we should think of a better name than 'pinned'.

eamodio commented:
For a name instead of `pinned` -- how about `preview` since I believe that's what its called in the docs (preview tab)?
Did you paste the wrong link (or am I missing something) because I thought this does address https://github.com/Microsoft/vscode/issues/10568.

As for exposing the information -- totally agree, I look into adding that. But for changing after the fact -- that feels less clear. Should you be able to go from something pinned to not pinned? Maybe so, but the UI currently doesn't provide a way for that direction (though an extension might have a good reason for it). What do you think?

Also can you elaborate on the last point about certain editors can't be pinned -- where would I find that, and would/should that be handled explicitly here?

jrieken commented:
Look here (https://github.com/Microsoft/vscode/issues/10568#issuecomment-242248377) which is about making the transition. Basically turning an existing editor into a pinned (from a preview editor)

> Also can you elaborate on the last point about certain editors can't be pinned

This is only an issue when the pinned-state is exposed as property. Similar to `viewColumn` is only applies to one the "big editors" not those of a peek view (find references) or the left of a diff editor

eamodio commented:
Ah got it -- going from unpinned --> pinned make perfect sense (was thinking that could just be handled by another call to `showTextDocument`, but something more explicit is probably best). Think it should be a 1-way deal, or also support going from pinned --> unpinned?

eamodio commented:
I made the first set of changes -- didn't rename pinned yet until there is agreement on a new name.
Also haven't looked into adding pinned as a property yet.
Since we moved column into the options, I added column support to the `diff` command too.

eamodio commented on src/vs/vscode.d.ts:
```diff
@@ -699,6 +699,26 @@ declare module 'vscode' {
 	/**
+	 * Represents options to configure the behavior of showing a [document](#TextDocument) in an [editor](#TextEditor).
+	 */
+	export interface ShowTextDocumentOptions {
```
@jrieken I don't love this name, let me know what you think

jrieken replied:
Maybe turning it around `TextDocumentShowOptions` or `TextDocumentViewOptions`?

eamodio replied:
`TextDocumentDisplayOptions`?

jrieken replied:
not sure yet.

eamodio commented on src/vs/workbench/api/node/mainThreadEditors.ts:
```diff
@@ -102,11 +104,16 @@ export class MainThreadEditors extends MainThreadEditorsShape {
 
-	$tryShowTextDocument(resource: URI, position: EditorPosition, preserveFocus: boolean): TPromise<string> {
+	$tryShowTextDocument(resource: URI, options: vscode.ShowTextDocumentOptions): TPromise<string> {
+		const position: EditorPosition = TypeConverters.fromViewColumn(options.column);
```
@jrieken I moved some of the data massaging into here, so that it could take the same `ShowTextDocumentOptions` (since I think you can technically call the proxy methods from inside an extension).

jrieken replied:
Don't use the TypeConverter here but send proper data. Do all the massaging before calling a main-side method. Rule of thumb is that `mainXYZ` files shouldn't depend on `extHostXZY` files

eamodio replied:
The main issue with that, is then we would need 2 "versions" of the `TextDocument[Show|View|Display]Options` -- one for the extensions with a `column` and one for the main-side with a `position`. Which is certainly doable, I was just trying to not have 2 types with almost the same fields. Are you fine with adding 2 types? And if so, where should the main-side one live and how should it be named to not cause confusion?

jrieken replied:
Two - one is API the other the DTO

jrieken commented:
@eamodio Can you go ahead and resolve the merge conflicts?

eamodio commented:
@jrieken is a rebase/force push fine to resolve the conflicts or should I do it in a follow-on commit?

jrieken commented:
> rebase/force push 

That's OK for me

eamodio commented:
@jrieken Are you good with `preview` as the name rather than pinned (even though it is the inverse)?

Also see comments above

jrieken commented:
Yeah, let go with `preview` for now

eamodio commented:
@jrieken I've resolved the conflicts and made all the code review changes above

jrieken commented on src/vs/vscode.d.ts:
```diff
@@ -699,6 +699,26 @@ declare module 'vscode' {
+		/**
+		 * An optional view column in which the [editor](#TextEditor) should be shown. The default is the [one](#ViewColumn.One), other values are adjusted to be __Min(column, columnCount + 1)__.
+		 */
+		column ?: ViewColumn,
```
Wow - your commit should have been blocked because that's a formatting issue...

eamodio replied:
Fixed

jrieken commented on src/vs/workbench/electron-browser/commands.ts:
```diff
@@ -373,15 +376,31 @@ export function registerCommands(): void {
 	});
 
-	CommandsRegistry.registerCommand('_workbench.diff', function (accessor: ServicesAccessor, args: [URI, URI, string, string]) {
+	CommandsRegistry.registerCommand('_workbench.diff', function (accessor: ServicesAccessor, args: [URI, URI, string, string, vscode.TextDocumentShowOptions]) {
```
it's dangerous to use the vscode-module here (actually every outside the extension host) because these type actually don't exist. for interfaces it's OK because they disappear but using `IEditorOption` would have been better

eamodio replied:
I don't see an `IEditorOption` -- there is `IEditorOptions`, but that interface doesn't have any of the fields. Are you suggesting adding them there? Or am I missing something?

jrieken commented:
Trying to understand why the test have failed...

jrieken commented on src/vs/workbench/api/node/extHostTextEditors.ts:
```diff
@@ -55,8 +55,26 @@ export class ExtHostEditors extends ExtHostEditorsShape {
+				pinned: true
+			};
+		} else {
+			options = {
```
The `else` cannot be unconditional because it will be hit when no second parameter is given. Add a check like `typeof columnOrOptions === 'object'`

eamodio replied:
Fixed

eamodio commented:
I think the fall-through case was causing the test failures -- they all passed now! :)

jrieken commented:
Merging now. Thanks for this PR, for reacting to feedback.

eamodio commented:
@jrieken thank you!

eamodio commented:
Look great -- just had that 1 comment on a possible behavior change.
